Cython/Compiler/Optimize.py

```diff
@@ -2580,26 +2581,56 @@ def _handle_simple_function_set(self, node, function, pos_args):
                 is_temp=node.is_temp,
                 py_name="set"))
 
-    PyFrozenSet_New_func_type = PyrexTypes.CFuncType(
-        Builtin.frozenset_type, [
-            PyrexTypes.CFuncTypeArg("it", PyrexTypes.py_object_type, None)
-        ])
-
+    # SimpleCallNode is replaced
     def _handle_simple_function_frozenset(self, node, function, pos_args):
+        print("==== _handle_simple_function_frozenset view")
+        print(node)
+        print(node.__dict__)
         if not pos_args:
-            pos_args = [ExprNodes.NullNode(node.pos)]
-        elif len(pos_args) > 1:
+            # pos_args = [ExprNodes.NullNode(node.pos)]
+            # print("!! Returning empty frozenset")
+            result = ExprNodes.FrozenSetNode(node.pos, is_temp=False, args=None)
+            result.is_literal = True
+            return result
+
+        # frozenset() expects no arguments, or one iterable argument
+        # https://docs.python.org/3.9/library/stdtypes.html#frozenset
+        if len(pos_args) > 1:
             return node
-        elif pos_args[0].type is Builtin.frozenset_type and not pos_args[0].may_be_none():
+
+
+        if pos_args[0].type is Builtin.frozenset_type :
+            # and not pos_args[0].may_be_none()
+            # print(pos_args[0].may_be_none())
+            # DOUBT: when do we need to care about may_be_none()
```

Comment:
Read what it says: if the argument is of type `frozenset` and cannot be `None`, then we know that it's a `frozenset` object and we can return it instead of passing it through the `frozenset()` constructor again. `frozenset(None)` would give you a runtime error.

Reply:
Search for `may_be_none(self)` in `ExprNodes.py`. It's one of the many generic methods that a syntax tree node can implement to expose a specific behaviour for a part of an expression (specifically, whether this expression can return a `None` value). There's some documentation in the `Node` and `ExprNode` classes.